In [1]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt

Загрузим данные и посмотрим на них.

In [2]:
df = pd.read_csv('/content/drive/MyDrive/Тинькофф.Поколение/train.csv', encoding='ISO-8859-1')

In [3]:
df.head()

Id  ... Rating
0   0  ...   80.0
1   1  ...  100.0
2   2  ...   71.0
3   3  ...   86.0
4   4  ...   86.0

[5 rows x 5 columns]

In [4]:
np.unique(df['Hotel_name']).shape # название отеля - категориальный признак

(243,)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2351 entries, 0 to 2350
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Id            2351 non-null   int64  
 1   Hotel_name    2351 non-null   object 
 2   Review_Title  2136 non-null   object 
 3   Review_Text   2351 non-null   object 
 4   Rating        2351 non-null   float64
dtypes: float64(1), int64(1), object(3)
memory usage: 92.0+ KB


Мы решаем задачу регрессии. Метрика - MSE/RMSE.

# Preprocessing.

Разберемся с NaN.

In [6]:
df.isna().sum()

Id                0
Hotel_name        0
Review_Title    215
Review_Text       0
Rating            0
dtype: int64

In [7]:
df['Review_Title'].fillna('no description', inplace=True) # заполним Nan'ы

Выкинем столбец Id, он бесполезен.

In [8]:
df = df.drop('Id', axis=1)

## Обработаем текстовые признаки, feature engineering.

### Первичная обработка: lower-case и убираем знаки препинания.

In [9]:
import re
df['Review_Text'] = df['Review_Text'].apply(lambda x: ' '.join(re.findall(r'\w+', x.lower())))
df['Review_Title'] = df['Review_Title'].apply(lambda x: ' '.join(re.findall(r'\w+', x.lower())))

In [10]:
df.head()

Hotel_name  ... Rating
0         Park Hyatt  ...   80.0
1     Hilton Chennai  ...  100.0
2  The Royal Regency  ...   71.0
3             Rivera  ...   86.0
4         Park Hyatt  ...   86.0

[5 rows x 4 columns]

### Обработка слов в текстовых признаках.

Нам предоставлены текстовые признаки на английском языке. С ними хорошо справится лемматизатор.

In [11]:
# Позаимствуем функции для лемматизации текста из семинарского ноутбука;)
from nltk import wordnet, pos_tag

def get_wordnet_pos(treebank_tag):
    my_switch = {
        'J': wordnet.wordnet.ADJ,
        'V': wordnet.wordnet.VERB,
        'N': wordnet.wordnet.NOUN,
        'R': wordnet.wordnet.ADV,
    }
    for key, item in my_switch.items():
        if treebank_tag.startswith(key):
            return item
    return wordnet.wordnet.NOUN

In [12]:
import nltk

nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/wordnet.zip.


True

In [13]:
from nltk import WordNetLemmatizer

def my_lemmatizer(sent):
    lemmatizer = WordNetLemmatizer()
    tokenized_sent = sent.split()
    pos_tagged = [(word, get_wordnet_pos(tag))
                 for word, tag in pos_tag(tokenized_sent)]
    return ' '.join([lemmatizer.lemmatize(word, tag)
                    for word, tag in pos_tagged])

Лемматизация.

In [14]:
df['Review_Text'] = df['Review_Text'].apply(my_lemmatizer) # лемматизируем и тексты, и заголовки
df['Review_Title'] = df['Review_Title'].apply(my_lemmatizer)

In [15]:
df.head()

Hotel_name  ... Rating
0         Park Hyatt  ...   80.0
1     Hilton Chennai  ...  100.0
2  The Royal Regency  ...   71.0
3             Rivera  ...   86.0
4         Park Hyatt  ...   86.0

[5 rows x 4 columns]

### Очищаем данные от stopwords.

Удалим филлерные слова, они не дают нам никакой информации и только мешают.

In [16]:
import nltk
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [17]:
from nltk.corpus import stopwords
sw_eng = set(stopwords.words('english'))

In [18]:
# выкидываем все филлеры
df['Review_Text'] = df['Review_Text'].apply(lambda x: ' '.join([word for word in x.split() if not word in sw_eng]))
df['Review_Title'] = df['Review_Title'].apply(lambda x: ' '.join([word for word in x.split() if not word in sw_eng]))

In [19]:
df.head()

Hotel_name  ... Rating
0         Park Hyatt  ...   80.0
1     Hilton Chennai  ...  100.0
2  The Royal Regency  ...   71.0
3             Rivera  ...   86.0
4         Park Hyatt  ...   86.0

[5 rows x 4 columns]

Well Done! Наконец-то нормальный читаемый текст!

### Подсчет pos/neg слов.

Интересно и полезно было бы посчитать количество слов с позитивной/негативной окраской в каждом тексте и получить новые признаки. Тем более нам предоставлены файлы - словари.

In [20]:
# Напишем функции для подсчета количества положительных и негативных слов
# ищем вхождения в предоставленные файлы
def count_pos(text):
  cnt_pos = 0
  for word in re.findall(r'[^,.;\s]+', text.lower()): # парсим данные регулярными выражениями
    with open('/content/drive/MyDrive/Тинькофф.Поколение/positive-words.txt') as f:
      if word in f.read().split():
        cnt_pos += 1
  return cnt_pos

def count_neg(text):
  cnt_neg = 0
  for word in re.findall(r'[^,.;\s]+', text.lower()):
    with open('/content/drive/MyDrive/Тинькофф.Поколение/negative-words.txt', encoding="ISO-8859-1") as g:
      if word in g.read().split():
        cnt_neg += 1

  return cnt_neg

In [21]:
df['Pos_Words'] = df['Review_Text'].apply(count_pos) # Введем новые признаки - количество позитивных и негативных слов
df['Neg_Words'] = df['Review_Text'].apply(count_neg)

In [22]:
df.head()

Hotel_name                        Review_Title  ... Pos_Words  Neg_Words
0         Park Hyatt                      refuge chennai  ...         4          0
1     Hilton Chennai                      hilton chennai  ...         4          0
2  The Royal Regency  worth rating show website price ok  ...         8          1
3             Rivera                           good stay  ...         4          2
4         Park Hyatt                    need improvement  ...         2          0

[5 rows x 6 columns]

## Представление текста.

In [23]:
y = df['Rating']

### Count Vectorizer.

In [24]:
from sklearn.feature_extraction.text import CountVectorizer

cv = CountVectorizer()
text_cv = cv.fit_transform(df['Review_Text'])
title_cv = cv.fit_transform(df['Review_Title'])

In [25]:
df_cv = df.drop(['Review_Title', 'Review_Text', 'Rating'], axis=1)

In [26]:
df_cv = pd.concat([df_cv, pd.DataFrame(text_cv.toarray())], axis=1) # добавляем эмбеддинги в конец датасета
df_cv = pd.concat([df_cv, pd.DataFrame(title_cv.toarray())], axis=1)
df_cv.columns = np.arange(df_cv.shape[1])

In [27]:
df_cv.head()

0     1     2     3     4     ...  4567  4568  4569  4570  4571
0         Park Hyatt     4     0     0     0  ...     0     0     0     0     0
1     Hilton Chennai     4     0     0     0  ...     0     0     0     0     0
2  The Royal Regency     8     1     0     0  ...     0     0     0     0     0
3             Rivera     4     2     0     0  ...     0     0     0     0     0
4         Park Hyatt     2     0     0     0  ...     0     0     0     0     0

[5 rows x 4572 columns]

### TF-IDF.

In [28]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf = TfidfVectorizer()
text_tf = tfidf.fit_transform(df['Review_Text'])
title_tf = tfidf.fit_transform(df['Review_Title'])

In [29]:
df_tf = df.drop(['Review_Title', 'Review_Text', 'Rating'], axis=1)

In [30]:
df_tf = pd.concat([df_tf, pd.DataFrame(text_tf.toarray())], axis=1) # добавляем эмбеддинги в конец датасета
df_tf = pd.concat([df_tf, pd.DataFrame(title_tf.toarray())], axis=1)
df_tf.columns = np.arange(df_tf.shape[1])

In [31]:
df_tf.head()

0     1     2     3     4     ...  4567  4568  4569  4570  4571
0         Park Hyatt     4     0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0
1     Hilton Chennai     4     0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0
2  The Royal Regency     8     1   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0
3             Rivera     4     2   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0
4         Park Hyatt     2     0   0.0   0.0  ...   0.0   0.0   0.0   0.0   0.0

[5 rows x 4572 columns]

### Word2Vec

In [32]:
import io
from tqdm import tqdm
from itertools import islice

def load_vectors(fname, limit): # воспользуемся функциями из конспектов:)
    fin = io.open(fname, 'r', encoding='utf-8', newline='\n', errors='ignore')
    n, d = map(int, fin.readline().split())
    data = {}
    for line in tqdm(islice(fin, limit), total=limit):
        tokens = line.rstrip().split(' ')
        data[tokens[0]] = np.array(list(map(float, tokens[1:])))
    return data

In [33]:
def text2vec(text):
    words = text.split()
    if len(words) == 0:
      words = ['no', 'description']
      
    return sum(list(map(lambda w: np.array(list(vecs.get(w, zero))), words))) / len(words)

In [34]:
vecs = load_vectors('/content/drive/MyDrive/Тинькофф.Поколение/crawl-300d-2M.vec', 2000000)
zero = sum(vecs.values()) / len(vecs)

100%|█████████▉| 1999995/2000000 [03:39<00:00, 9094.39it/s] 


In [93]:
text_w = np.array([list(text2vec(row[1]['Review_Text'])) for row in df.iterrows()]) # получаем массив эмбеддингов
title_w = np.array([list(text2vec(row[1]['Review_Title'])) for row in df.iterrows()])

In [94]:
df_w = df.drop(['Review_Title', 'Review_Text', 'Rating'], axis=1)

In [95]:
df_w = pd.concat([df_w, pd.DataFrame(text_w)], axis=1) # добавляем эмбеддинги в конец датасета
df_w = pd.concat([df_w, pd.DataFrame(title_w)], axis=1)
df_w.columns = np.arange(df_w.shape[1])

In [38]:
df_w.head()

0    1    2         3    ...       599      600       601       602
0         Park Hyatt    4    0 -0.014004  ... -0.205800 -0.28315 -0.043350 -0.003450
1     Hilton Chennai    4    0  0.017690  ... -0.179050 -0.10455 -0.100100 -0.102750
2  The Royal Regency    8    1 -0.072474  ... -0.076383  0.07920 -0.042883  0.001283
3             Rivera    4    2 -0.010500  ...  0.016700  0.08585 -0.145850  0.025300
4         Park Hyatt    2    0  0.001365  ... -0.042850 -0.19565  0.194100 -0.185250

[5 rows x 603 columns]

# Модель, обучение и предсказание.

In [39]:
!pip install catboost # установим катбуст

     |████████████████████████████████| 67.3MB 101kB/s 


In [40]:
from sklearn.model_selection import train_test_split
 # все 3 полученных (для каждого представления текстов) датасета сплитим
X_train_cv, X_test_cv, y_train_cv, y_test_cv = train_test_split(df_cv, y, test_size=0.2, random_state=42)
X_train_tf, X_test_tf, y_train_tf, y_test_tf = train_test_split(df_tf, y, test_size=0.2, random_state=42)
X_train_w, X_test_w, y_train_w, y_test_w = train_test_split(df_w, y, test_size=0.2, random_state=42)

## CatBoost.

Count Vectorizer.

In [41]:
from catboost import CatBoostRegressor
from sklearn.metrics import mean_squared_error

cbr = CatBoostRegressor(cat_features=[0], random_seed=42, verbose=False)
cbr.fit(X_train_cv, y_train_cv)
y_pred_cv = cbr.predict(X_test_cv)
print('MSE: {}'.format(mean_squared_error(y_test_cv, y_pred_cv)))
print('RMSE: {}'.format(mean_squared_error(y_test_cv, y_pred_cv, squared=False)))

MSE: 199.21628520848265
RMSE: 14.114399923782896


TF-IDF.

In [42]:
cbr = CatBoostRegressor(cat_features=[0], random_seed=42, verbose=False)
cbr.fit(X_train_tf, y_train_tf)
y_pred_tf = cbr.predict(X_test_tf)
print('MSE: {}'.format(mean_squared_error(y_test_tf, y_pred_tf)))
print('RMSE: {}'.format(mean_squared_error(y_test_tf, y_pred_tf, squared=False)))

MSE: 183.27807499645013
RMSE: 13.538023304620587


Word2Vec.

In [43]:
cbr = CatBoostRegressor(cat_features=[0], random_seed=42, verbose=False)
cbr.fit(X_train_w, y_train_w)
y_pred_w = cbr.predict(X_test_w)
print('MSE: {}'.format(mean_squared_error(y_test_w, y_pred_w)))
print('RMSE: {}'.format(mean_squared_error(y_test_w, y_pred_w, squared=False)))

MSE: 175.51253614423837
RMSE: 13.24811443731667


В статьях на Медиуме (например, в [этой](https://medium.com/@bedigunjit/simple-guide-to-text-classification-nlp-using-svm-and-naive-bayes-with-python-421db3a72d34)) для nlp classification активно используются такие модели, как Naive Bayes и SVM. Попробуем для задачи регрессии использовать соответствующие регрессоры.

In [45]:
df_cv_enc = pd.get_dummies(df_cv)
df_tf_enc = pd.get_dummies(df_tf)
df_w_enc = pd.get_dummies(df_w)

In [46]:
X_train_cv_enc, X_test_cv_enc, y_train_cv_enc, y_test_cv_enc = train_test_split(df_cv_enc, y, test_size=0.2, random_state=42)
X_train_tf_enc, X_test_tf_enc, y_train_tf_enc, y_test_tf_enc = train_test_split(df_tf_enc, y, test_size=0.2, random_state=42)
X_train_w_enc, X_test_w_enc, y_train_w_enc, y_test_w_enc = train_test_split(df_w_enc, y, test_size=0.2, random_state=42)

## Bayesian Ridge, Support Vector Regression, SGD Classifier.

Count Vectorizer.

In [56]:
from sklearn.linear_model import BayesianRidge
from sklearn.svm import SVR
from sklearn.linear_model import SGDClassifier

for name in [BayesianRidge, SVR, SGDClassifier]:
  model = name()
  print(model)
  model.fit(X_train_cv_enc, y_train_cv_enc)
  y_pred_cv_enc = model.predict(X_test_cv_enc)
  print('\nMSE: {}'.format(mean_squared_error(y_test_cv_enc, y_pred_cv_enc)))
  print('RMSE: {}\n'.format(mean_squared_error(y_test_cv_enc, y_pred_cv_enc, squared=False)))

BayesianRidge(alpha_1=1e-06, alpha_2=1e-06, alpha_init=None,
              compute_score=False, copy_X=True, fit_intercept=True,
              lambda_1=1e-06, lambda_2=1e-06, lambda_init=None, n_iter=300,
              normalize=False, tol=0.001, verbose=False)

MSE: 216.89620894870026
RMSE: 14.727396543472993

SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

MSE: 314.76528867742667
RMSE: 17.741625874688786

SGDClassifier(alpha=0.0001, average=False, class_weight=None,
              early_stopping=False, epsilon=0.1, eta0=0.0, fit_intercept=True,
              l1_ratio=0.15, learning_rate='optimal', loss='hinge',
              max_iter=1000, n_iter_no_change=5, n_jobs=None, penalty='l2',
              power_t=0.5, random_state=None, shuffle=True, tol=0.001,
              validation_fraction=0.1, verbose=0, warm_start=False)

MSE: 279.86411889596604
RMSE: 16.729139813390468



TF-IDF.

In [54]:
for name in [BayesianRidge, SVR, SGDClassifier]:
  model = name()
  print(model)
  model.fit(X_train_tf_enc, y_train_tf_enc)
  y_pred_tf_enc = model.predict(X_test_tf_enc)
  print('\nMSE: {}'.format(mean_squared_error(y_test_tf_enc, y_pred_tf_enc)))
  print('RMSE: {}\n'.format(mean_squared_error(y_test_tf_enc, y_pred_tf_enc, squared=False)))

BayesianRidge(alpha_1=1e-06, alpha_2=1e-06, alpha_init=None,
              compute_score=False, copy_X=True, fit_intercept=True,
              lambda_1=1e-06, lambda_2=1e-06, lambda_init=None, n_iter=300,
              normalize=False, tol=0.001, verbose=False)

MSE: 196.209815868169
RMSE: 14.007491419528659

SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

MSE: 324.92088437047005
RMSE: 18.025561971003015

SGDClassifier(alpha=0.0001, average=False, class_weight=None,
              early_stopping=False, epsilon=0.1, eta0=0.0, fit_intercept=True,
              l1_ratio=0.15, learning_rate='optimal', loss='hinge',
              max_iter=1000, n_iter_no_change=5, n_jobs=None, penalty='l2',
              power_t=0.5, random_state=None, shuffle=True, tol=0.001,
              validation_fraction=0.1, verbose=0, warm_start=False)

MSE: 270.7515923566879
RMSE: 16.454531058547



Word2Vec.

In [57]:
for name in [BayesianRidge, SVR, SGDClassifier]:
  model = name()
  print(model)
  model.fit(X_train_w_enc, y_train_w_enc)
  y_pred_w_enc = model.predict(X_test_w_enc)
  print('\nMSE: {}'.format(mean_squared_error(y_test_w_enc, y_pred_w_enc)))
  print('RMSE: {}\n'.format(mean_squared_error(y_test_w_enc, y_pred_w_enc, squared=False)))

BayesianRidge(alpha_1=1e-06, alpha_2=1e-06, alpha_init=None,
              compute_score=False, copy_X=True, fit_intercept=True,
              lambda_1=1e-06, lambda_2=1e-06, lambda_init=None, n_iter=300,
              normalize=False, tol=0.001, verbose=False)

MSE: 223.71053429252595
RMSE: 14.95695605036419

SVR(C=1.0, cache_size=200, coef0=0.0, degree=3, epsilon=0.1, gamma='scale',
    kernel='rbf', max_iter=-1, shrinking=True, tol=0.001, verbose=False)

MSE: 309.2319401542456
RMSE: 17.58499190088655

SGDClassifier(alpha=0.0001, average=False, class_weight=None,
              early_stopping=False, epsilon=0.1, eta0=0.0, fit_intercept=True,
              l1_ratio=0.15, learning_rate='optimal', loss='hinge',
              max_iter=1000, n_iter_no_change=5, n_jobs=None, penalty='l2',
              power_t=0.5, random_state=None, shuffle=True, tol=0.001,
              validation_fraction=0.1, verbose=0, warm_start=False)

MSE: 315.9447983014862
RMSE: 17.774836097739023



# Заключительная часть.

## Модель.

После всех проведенных испытаний стало ясно, что лучше всех справилась модель `CatBoostRegressor` со всеми видами представления данных. Немного хуже сработал `BayesianRidge`, о `SVR` и `SGDClassifier` лучше умолчать:)

## Представление текстовых данных.

`CatBoostRegressor` показал лучший результат при работе с `Word2Vec`. `BayesianRidge` лучше всего сработал, наоборот, с `TF-IDF`, а с `Word2Vec` показал худший результат. Остальные модели далеко позади.

## Вывод.

Таким образом, фаворит сегодняшнего вечера - `CatBoostRegressor` с представлением текстовых данных с помощью `Word2Vec`.

Результат:

MSE: 175.51253614423837

RMSE: 13.24811443731667

## Предсказание на тестовой части.

In [103]:
df_test = pd.read_csv('/content/drive/MyDrive/Тинькофф.Поколение/test.csv', encoding='ISO-8859-1')

In [104]:
id_col = df_test['Id']

In [105]:
df_test.isna().sum() # проверим на Nan'ы

Id                0
Hotel_name        0
Review_Title    209
Review_Text       0
dtype: int64

In [78]:
df_test['Review_Title'].fillna('no description', inplace=True) # заполним Nan'ы

In [79]:
df_test['Review_Text'] = df_test['Review_Text'].apply(lambda x: ' '.join(re.findall(r'\w+', x.lower()))) # lower-case, очищаем от знаков
df_test['Review_Title'] = df_test['Review_Title'].apply(lambda x: ' '.join(re.findall(r'\w+', x.lower())))

In [80]:
df_test['Review_Text'] = df_test['Review_Text'].apply(my_lemmatizer) # лемматизируем и тексты, и заголовки
df_test['Review_Title'] = df_test['Review_Title'].apply(my_lemmatizer)

In [81]:
# выкидываем все филлеры
df_test['Review_Text'] = df_test['Review_Text'].apply(lambda x: ' '.join([word for word in x.split() if not word in sw_eng]))
df_test['Review_Title'] = df_test['Review_Title'].apply(lambda x: ' '.join([word for word in x.split() if not word in sw_eng]))

In [82]:
df_test['Pos_Words'] = df_test['Review_Text'].apply(count_pos) # Введем новые признаки - количество позитивных и негативных слов
df_test['Neg_Words'] = df_test['Review_Text'].apply(count_neg)

In [83]:
text_w = np.array([list(text2vec(row[1]['Review_Text'])) for row in df_test.iterrows()]) # получаем массив эмбеддингов
title_w = np.array([list(text2vec(row[1]['Review_Title'])) for row in df_test.iterrows()])

In [84]:
df_test = df_test.drop(['Review_Title', 'Review_Text', 'Id', 'Hotel_name'], axis=1) # названия отелей нам, впринципе, и не нужны
df_w = df_w.drop([0], axis=1)

In [85]:
df_test = pd.concat([df_test, pd.DataFrame(text_w)], axis=1) # добавляем эмбеддинги в конец датасета
df_test = pd.concat([df_test, pd.DataFrame(title_w)], axis=1)
df_test.columns = np.arange(df_test.shape[1])
df_w.columns = np.arange(df_w.shape[1])

In [86]:
df_test.head()

0    1         2         3    ...      598      599      600       601
0    5    3 -0.085957 -0.088300  ... -0.10710 -0.21510 -0.24780  0.419850
1    2    0  0.021571  0.018029  ... -0.02170  0.33580 -0.06960  0.149900
2    5    0 -0.015844 -0.110565  ... -0.08835  0.02435 -0.08605  0.037475
3    1    1  0.071125  0.078262  ...  0.04670  0.59165 -0.10840  0.038400
4   10    0  0.022224  0.019510  ... -0.10144  0.21420 -0.06212 -0.039280

[5 rows x 602 columns]

In [102]:
cbr = CatBoostRegressor(random_seed=42, verbose=False) # обучение на лучшей модели
cbr.fit(df_w, y)
y_pred = cbr.predict(df_test)

In [106]:
submit = pd.DataFrame() # предсказание на тесте
submit['Id'] = id_col
submit['Rating'] = y_pred
submit.to_csv('submit.csv', index = False)